File: src/app/core/service/inventory.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from "@angular/common/http";
import {Inventory} from "../model/inventory.model";
import {Observable} from "rxjs";

@Injectable({
  providedIn: 'root'
})
export class InventoryService {

  domain: string = "http://localhost:8080/inventory";

  constructor(private http: HttpClient) {
  }


  get(pageIndex: number, pageSize: number, minPrice: number, maxPrice: number, minMileage: number, maxMileage: number, transmissionToggle: string[], driveToggle: string[], fuelToggle: string[], dealershipToggle: string[], sortBy: string, sortDir: string, make: string): Observable<Inventory[]> {
    let params = new HttpParams();
    params = params.append('page', pageIndex);
    params = params.append('size', pageSize);
    params = params.append('minPrice', minPrice);
    params = params.append('maxPrice', maxPrice);
    params = params.append('minMileage', minMileage);
    params = params.append('maxMileage', maxMileage);
    params = params.append('transmissionToggle', transmissionToggle.toString());
    params = params.append('driveToggle', driveToggle.toString());
    params = params.append('fuelToggle', fuelToggle.toString());
    params = params.append('dealershipToggle', dealershipToggle.toString());
    params = params.append('sortBy', sortBy);
    params = params.append('sortDir', sortDir);
    params = params.append('make', make);
    return this.http.get<Inventory[]>(this.domain, {params: params});
  }

  getMakes(): Observable<string[]> {
    return this.http.get<string[]>(this.domain + "/makes");
  }

  getById(id: string): Observable<Inventory> {
    return this.http.get<Inventory>(this.domain + "/" + id);
  }

  post(data: Inventory): Observable<Inventory> {
    return this.http.post<Inventory>(this.domain, data);
  }

  update(id: string, data: Inventory): Observable<Inventory> {
    return this.http.post<Inventory>(this.domain + "/" + id, data)
  }

  delete(id: string): Observable<unknown> {
    return this.http.delete(this.domain + "/" + id);
  }

}

File: src/main/java/io/infsci2710/autoapi/inventory/InventoryService.java
package io.infsci2710.autoapi.inventory;

import io.infsci2710.autoapi.sales.Sales;
import lombok.AllArgsConstructor;
import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Objects;

@Service
@AllArgsConstructor
public class InventoryService {

    @Autowired
    InventoryRepository repository;

    public Mono<Page<Inventory>> findAllBy(Pageable pageable, int minPrice, int maxPrice, int minMileage, int maxMileage, String transmissionToggle, String driveToggle, String fuelToggle, String dealershipToggle) {
        List<String> transmissionToggleArr = Arrays.asList(transmissionToggle.split(","));
        List<String> driveToggleArr = Arrays.asList(driveToggle.split(","));
        List<String> fuelToggleArr = Arrays.asList(fuelToggle.split(","));
        List<String> dealershipToggleArr = Arrays.asList(dealershipToggle.split(","));
        return repository.findAllByPriceBetweenAndMileageBetweenAndTransmissionInAndDriveInAndFuelInAndDealershipIdIn(minPrice, maxPrice, minMileage, maxMileage, transmissionToggleArr, driveToggleArr, fuelToggleArr, dealershipToggleArr, pageable)
                .collectList()
                .zipWith(this.repository.findAllByPriceBetweenAndMileageBetweenAndTransmissionInAndDriveInAndFuelInAndDealershipIdIn(minPrice, maxPrice, minMileage, maxMileage, transmissionToggleArr, driveToggleArr, fuelToggleArr, dealershipToggleArr,null).count())
                .map(t -> new PageImpl<>(t.getT1(), pageable, t.getT2()));
    }

    public Mono<Page<Inventory>> findAllByMake(Pageable pageable, int minPrice, int maxPrice, int minMileage, int maxMileage, String transmissionToggle, String driveToggle, String fuelToggle, String dealershipToggle, String make) {
        List<String> transmissionToggleArr = Arrays.asList(transmissionToggle.split(","));
        List<String> driveToggleArr = Arrays.asList(driveToggle.split(","));
        List<String> fuelToggleArr = Arrays.asList(fuelToggle.split(","));
        List<String> dealershipToggleArr = Arrays.asList(dealershipToggle.split(","));
        return repository.findAllByPriceBetweenAndMileageBetweenAndTransmissionInAndDriveInAndFuelInAndDealershipIdInAndMakeEquals(minPrice, maxPrice, minMileage, maxMileage, transmissionToggleArr, driveToggleArr, fuelToggleArr, dealershipToggleArr,make, pageable)
                .collectList()
                .zipWith(this.repository.findAllByPriceBetweenAndMileageBetweenAndTransmissionInAndDriveInAndFuelInAndDealershipIdInAndMakeEquals(minPrice, maxPrice, minMileage, maxMileage, transmissionToggleArr, driveToggleArr, fuelToggleArr, dealershipToggleArr,make, null).count())
                .map(t -> new PageImpl<>(t.getT1(), pageable, t.getT2()));
    }

    public Flux<Object> getMakes() { return repository.getDistinctMakes().map(data -> data.getMake()); }

    public Mono<Inventory> save(Inventory tuple) { return repository.insertInventory(tuple.getVin(), tuple.getMake(), tuple.getModel(), tuple.getModel_year(), tuple.getEngine(), tuple.getMileage(), tuple.getPrice(), tuple.getColor(), tuple.getTransmission(), tuple.getDrive(), tuple.getFuel(), tuple.getDealershipId()); }

    public Mono<Inventory> update(String id,Inventory tuple) { return repository.save(tuple); }

    public Mono<Inventory> delete(final String id) {
        final Mono<Inventory> dbObj = getById(id);
        if (Objects.isNull(dbObj)) {
            return Mono.empty();
        }
        return getById(id).switchIfEmpty(Mono.empty()).filter(Objects::nonNull).flatMap(tupleToDelete ->
                repository.delete(tupleToDelete).then(Mono.just(tupleToDelete)));
    }

    public Mono<Inventory> getById(final String id) { return repository.findById(id).switchIfEmpty(Mono.empty()); }


}
